### PR TITLE
PR for Refactor shutdown sequence to use plugin.stop

### LINKED ISSUE #3: refactor shutdown sequence to use plugin.stop
Change shutdown sequence to be triggered by plugin.stop instead of ShutdownSignal exception.
Also remove any calls to: shutdown, finished, finished?, running? or terminating?
This depends on https://github.com/elastic/logstash/pull/3895

### PR DESCRIPTION
**Pull request by guyboertje**
Shutdown semantics: Add stop method and shared_example
Add general tests (none before)
slight refactor
tests pass

Notes for Reviewers:
- This is a mixed PR, it contains code changes for the fix and general refactoring, because this plugin is very distant from core, the codebase has changed little since 2012 and there were no tests.
- The `require`s are not lazy loaded.
- Following our change in understanding about ivars and concurrency, I moved the ivar creation into the `initialize` method.
- I feared that when multiple rooms were specified, all but the last MUC client might be GC'd so I stored them in an array.
- I made the client and muc_clients available as attr readers so the tests could mimic the block call action when receiving messages.
- The tests use Mock classes that provide the same API as the Jabber Client and MUCSimpleClient.  Normally a test purist would say this is wrong because the API could change etc.  I considered this and figured that as the xmpp4r gem development has essentially stopped the risk was v low.
- The Mock classes are inline and not in a separate file so others can read them easily. I suspect, beyond bug fixes in this PR, this plugin will probably not see any more changes for some time.

Closes #3

### CODE REVIEW

File: lib/logstash/inputs/xmpp.rb [hunk 1]
```diff
@@ -32,26 +36,32 @@ class LogStash::Inputs::Xmpp < LogStash::Inputs::Base
...
     @client.auth(@password.value)
     @client.send(Jabber::Presence.new.set_type(:available))
-
-    # load the MUC Client if we are joining rooms.
-    require 'xmpp4r/muc/helper/simplemucclient' if @rooms && !@rooms.empty?
   end # def register
 
-  public
+  # for testing access
+  attr_reader :client, :muc_clients
```
Comment (talevy) [inline lib/logstash/inputs/xmpp.rb]: should we move this to the top or bottom of the file for easy finding?

File: lib/logstash/inputs/xmpp.rb [hunk 2]
```diff
@@ -67,14 +80,32 @@ def run(queue)
...
+    # [GUY] the two clients: muc and client maintain their own threads
+    # and will call the blocks in those threads so this one should sleep
+    # this is how the xmpp4r examples do it (not saying it is correct)
     sleep
+
+    @client.close
   end # def run
 
+  def stop
+    super
```
Comment (talevy) [inline lib/logstash/inputs/xmpp.rb]: no need to call `super` per he latest changes which occurred on the `3895` logstash PR

File: spec/inputs/xmpp_spec.rb [hunk 3]
```diff
@@ -1 +1,106 @@
...
+      rooms.each do |room|
+        event = queue.shift
+        expect(event['message']).to eq('Hello Foo')
+        expect(event['from']).to eq('bar')
+        expect(event['room']).to eq(room)
+      end
+    end
+  end
+
+  context 'when shuting down' do
```
Comment (talevy) [inline spec/inputs/xmpp_spec.rb]: s/shuting/shutting

File: lib/logstash/inputs/xmpp.rb [hunk 4]
```diff
@@ -67,14 +80,32 @@ def run(queue)
...
+    @run_thread.run if @run_thread.alive?
```
Comment (talevy) [inline lib/logstash/inputs/xmpp.rb]: why is the running thread told to wake up and run? the client hijacks the thread and manages it on its own?

File: lib/logstash/inputs/xmpp.rb (hunk 2 above)
Comment (guyboertje) [inline lib/logstash/inputs/xmpp.rb]: OK

File: lib/logstash/inputs/xmpp.rb (hunk 4 above)
Comment (guyboertje) [inline lib/logstash/inputs/xmpp.rb]: The answer is in the comment before the sleep in `run` method.

File: spec/inputs/xmpp_spec.rb (hunk 3 above)
Comment (guyboertje) [inline spec/inputs/xmpp_spec.rb]: thx

File: lib/logstash/inputs/xmpp.rb (hunk 1 above)
Comment (guyboertje) [inline lib/logstash/inputs/xmpp.rb]: My thinking: as its used in methods from register downwards it should be above `def register`.

We don't really enforce the classic ruby file layout and for plugins I think the config DSL should standout at the top. My 2c.

File: lib/logstash/inputs/xmpp.rb [hunk 5]
```diff
@@ -59,22 +70,41 @@ def run(queue)
...
+    # [GUY] the two clients: muc and client maintain their own threads
+    # and will call the blocks in those threads so this one should sleep
+    # this is how the xmpp4r examples do it (not saying it is correct)
     sleep
+
+    client.close
   end # def run
 
+  def stop
+    super
```
Comment (jsvd) [inline lib/logstash/inputs/xmpp.rb]: super is no longer needed

File: spec/inputs/xmpp_spec.rb [hunk 6]
```diff
@@ -1 +1,106 @@
+# encoding: utf-8
 require "logstash/devutils/rspec/spec_helper"
+require 'logstash/inputs/xmpp'
+
+module Jabber
```
Comment (purbon) [inline spec/inputs/xmpp_spec.rb]: I would move this mocks to an `spec_helper.rb` file or other kind of support helper concept.

File: spec/inputs/xmpp_spec.rb [hunk 7]
```diff
@@ -1 +1,106 @@
...
+    allow(Jabber::MUC::SimpleMUCClient).to receive(:new) do |client|
+      Jabber::MucClientMock.new(client)
+    end
+  end
+
+  context 'when running' do
+    let(:queue) { [] }
+    let(:msg)   { 'Hello Logstash'}
+
+    subject { described_class.new(config) }
```
Comment (purbon) [inline spec/inputs/xmpp_spec.rb]: Even if the code is valid, I would recommend using `LogStash::Inputs::Xmpp` here, not sure everyone will grasp at this level what described_class actually is. what do you think?

File: spec/inputs/xmpp_spec.rb [hunk 8]
```diff
@@ -1 +1,106 @@
...
+
+    before { subject.register }
+
+    it 'pushes normal events into a queue' do
+      plugin_thread = Thread.new(subject, queue) { |subj, que| subj.run(que) }
```
Comment (purbon) [inline spec/inputs/xmpp_spec.rb:37]: Even if this might be valid, I would recommend to DRY this options out in a helper method or in a before section. Like this the complexity of the test examples would be less so easy to understand in a while :+1:

File: spec/inputs/xmpp_spec.rb (hunk 7 above)
Comment (guyboertje) [inline spec/inputs/xmpp_spec.rb]: I think it is standard rspec. I am torn, on the one hand I can see your point about spoon feeding some people, on the other - those people should be able to do a simple Google search for the answer.

File: spec/inputs/xmpp_spec.rb (hunk 6 above)
Comment (guyboertje) [inline spec/inputs/xmpp_spec.rb]: I considered this and made a point about it in the Reviewers Notes above.  What rule of thumb should we apply here? On the one hand it is tidy to move the mocks to their own file and on the other hand, moving them will make it less easy to read and follow the whole spec suite in one go.

I cannot find any solid guidance in the Internet about extract or not wrt helpers.  Even the [Rspec style guide](http://betterspecs.org/) is silent on this.

The only RSpec convention is to put them in /support and they get auto-loaded but then the same people that don't know what described_class means will not know about this RSpec convention either.

File: spec/inputs/xmpp_spec.rb (hunk 8 above)
Comment (guyboertje) [inline spec/inputs/xmpp_spec.rb:37]: I prefer DRY'ing to a before section, but I think it may look less easy to understand. I will try a spike.

@purbon: I like to think of the 'solar system' when I imagine the LS eco system.  Logstash core is the Sun and the plugins and support classes are planets.  In this analogy, this plugin is Pluto - seldom seen and not often focussed on.  I think we can get away with some not DRY stuff when the plugin is so outlying.

File: spec/inputs/xmpp_spec.rb (hunk 6 above)
Comment (purbon) [inline spec/inputs/xmpp_spec.rb]: Hi,
 as there no official logstash style guide about it, it end up very personal. But in my experience, keeping the test without the definition of mocks and other support classes makes sense as code is simpler and easier to understand then.

I would strongly recommend you to do it as will keep things easy to understand and manage in the long term. But again this is my opinion, i tend to like everything in their place, and for this example mocks belong to their definition files and test belong here, mixing them confuses the read, and somehow makes stuff tied up.

On the other side, even if not official this is a pattern I'm trying to enforce in the hole code base, so it would be nice to have mocks defined in helper files that then gets used in the examples.

Again, is the same example as having smoller methods as possible, we're humans and to understand big pieces of code it gets error prone.

File: spec/inputs/xmpp_spec.rb (hunk 7 above)
Comment (purbon) [inline spec/inputs/xmpp_spec.rb]: I know is standard and I know what I means, just found so many people confused with it, on the other side it requires you to remember what you defined in the upper describe section, so is also error prone in that point too.

Both options are ok and do their work, just a matter usage.

**Comment by purbon** [comment]
LGTM, keep in mind to squash the commits before merging :+1:
